src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts

```diff
@@ -1,79 +1,103 @@
-import { bufferToBigInt, BUFFER_EMPTY } from "../../utils";
-import { BaseJsonRpcType } from "./json-rpc-base-types";
-// TODO(perf): rewrite this stuff since it isn't really caching anything
+import { bufferToBigInt } from "../../utils/buffer-to-bigint";
+import { BaseJsonRpcType, JsonRpcInputArg } from "./json-rpc-base-types";
+const BUFFER_EMPTY = Buffer.alloc(0);
+
 export class Quantity extends BaseJsonRpcType {
+  private static DEFAULT_STRING_VALUE = "0x0";
+
   _nullable: boolean = false;
-  public static from(
-    value: number | bigint | string | Buffer,
-    nullable = false
-  ) {
+
+  public static from(value: JsonRpcInputArg, nullable = false) {
     if (value instanceof Quantity) return value;
-    const q = new Quantity(value);
-    q._nullable = nullable;
-    return q;
+    return new Quantity(value, nullable);
+  }
+
+  constructor(value: JsonRpcInputArg, nullable?: boolean) {
+    super(value);
+    if (typeof(value) === "string" && value === "0x") {
+      throw new Error("Cannot wrap a 0x value as a json-rpc Quantity type; Quantity must contain at least one digit");
+    }
+    this._nullable = nullable;
   }
+
   public toString(): string | null {
-    // TODO(perf): memoize this stuff
-    if (Buffer.isBuffer(this.value)) {
-      let val = this.value.toString("hex").replace(/^(?:0+(.+?))?$/, "$1");
-
-      if (val === "") {
-        if (this._nullable) {
-          return null;
-        }
-        // RPC Quantities must represent `0` as `0x0`
-        return "0x0";
-      }
-      return `0x${val}`;
-    } else if (this.value == null) {
-      return "0x";
-    } else {
-      return super.toString();
+    if (this.bufferValue == null) {
+      return this._nullable? this.bufferValue : Quantity.DEFAULT_STRING_VALUE;
     }
+
+    let firstNonZeroByte = 0;
+    for (firstNonZeroByte = 0; firstNonZeroByte < this.bufferValue.length; firstNonZeroByte++) {
+      if (this.bufferValue[firstNonZeroByte] !== 0) break;
+    }
+
+    let val = this.bufferValue.toString("hex", firstNonZeroByte);
+
+    if (val.length === 0 || val === "0") {
+      // RPC Quantities must represent `0` as `0x0`
+      return this._nullable ? null : Quantity.DEFAULT_STRING_VALUE;
+    }
+    if (val[0] === "0") {
```

Comment:
Can you explain this line? It isn't clear to me why we're doing this

Reply:
Yeah, that was wrong. I've fixed it up.

`Quantity.from(Buffer.alloc(0), true).toString()` would return `null` :/ which is now covered in this test: https://github.com/trufflesuite/ganache/blob/74a17afe9d084e888d4b4895690ab0af859b9013/src/packages/utils/tests/json-rpc-quantity.test.ts#L54